base/multidimensional.jl

```diff
@@ -226,6 +226,14 @@ index_shape_dim(A, dim, ::Colon) = (trailingsize(A, dim),)
 @inline index_shape_dim(A, dim, i::AbstractArray{Bool}, I...) = (sum(i), index_shape_dim(A, dim+1, I...)...)
 @inline index_shape_dim{N}(A, dim, i::AbstractArray{CartesianIndex{N}}, I...) = (size(i)..., index_shape_dim(A, dim+N, I...)...)
 
+@inline decolon(A::AbstractVector, ::Colon) = (indices(A,1),)
+@inline decolon(A::AbstractArray, ::Colon) = (1:length(A),)
+@inline decolon(A::AbstractArray, I...) = decolon_dim(A, 1, I...)
+@inline decolon_dim(A::AbstractArray, dim)                = ()
+@inline decolon_dim(A::AbstractArray, dim, ::Colon)       = dim == ndims(A) ?  (indices(A, dim),) : (1:trailingsize(A, dim),)
+@inline decolon_dim(A::AbstractArray, dim, i1, I...)      = (i1, decolon_dim(A, dim+1, I...)...)
+@inline decolon_dim(A::AbstractArray, dim, ::Colon, I...) = (indices(A, dim), decolon_dim(A, dim+1, I...)...)
```

Comment:
Yeah, I know.  It's extra messy because folks have also extended `to_index` in packages like DataArrays and NullableArrays.

But it could potentially happen alongside a deprecation that separates its behavior to `to_nonscalar_indexes`/`to_nonscalar_index`.  Probably worth doing separately.  This guy is more than big enough as is.